lib/util/errors.js

```diff
@@ -0,0 +1,61 @@
+'use strict';
+
+const {
+  Error,
+  SafeMap,
+} = primordials;
+
+const {
+  AbortError,
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_ERROR_MESSAGE,
+    ERR_MISSING_ERROR_MESSAGE,
+  },
+  makeErrorWithCode: _makeErrorWithCode,
+} = require('internal/errors');
+
+const {
+  DOMException,
+} = internalBinding('messaging');
+
+const {
+  validateFunction,
+  validateObject,
+  validateString,
+} = require('internal/validators');
+
+const {
+  isMap,
+} = require('util/types');
+
+function makeErrorWithCode(code, options = {}) {
+  validateString(code, 'code');
+  validateObject(options, 'options');
+  const {
+    Base = Error,
+    messages = new SafeMap(),
+  } = options;
+  validateFunction(Base, 'options.Base');
+  if (!isMap(messages)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'options.messages',
+      'Map',
+      messages);
+  }
+  const message = messages.get(code);
+  if (message === undefined) {
+    throw new ERR_MISSING_ERROR_MESSAGE(code);
+  }
+  if (typeof message !== 'function' &&
+      typeof message !== 'string') {
+    throw new ERR_INVALID_ERROR_MESSAGE(code);
```

Comment:
Why not ERR_INVALID_ARG_TYPE ? It supports multiple types and is used almost everywhere else for this kind of error.

Reply:
I realize `message` is not an arg here. Why is it only validated now rather than before adding it to the messages map?

Reply:
Please let me think a bit about the proposed API. There's something that doesn't feel right.

Reply:
The API was designed with minimal changes to the internals in mind.